src/utils.py

```diff
@@ -98,7 +98,11 @@ def buffer_if_necessary(shape):
     Following the advice given here:
     https://github.com/Toblerity/Shapely/issues/344
     """
-    if not shape.is_valid:
-        shape = shape.buffer(0.0)
-    assert shape.is_valid
-    return shape
\ No newline at end of file
+    if shape.is_valid:
+        return shape
+
+    new_shape = shape.buffer(0.0)
+    assert new_shape.is_valid
+    assert np.isclose(new_shape.area, new_shape.area)
```

Comment:
Useful addition!

Would be good to know the tolerance here. Is this close to the squaremeter, to the square kilometer, or to anything else?

Reply:
It's in whatever coordinate system the shapes are in, which is frequently WSG84 (i.e. the values are meaningless). It doesn't seem worth the penalty to update the CRS to being metric just for this.

Reply:
I agree. But even when it's given in WSG84 we can give that information. We can give an upper bound for sure. And the order of magnitude would be interesting here.

Reply:
`np.isclose` defaults to `rtol=1e-05, atol=1e-08`, which seems to be quite a tight tolerance irrespective of coordinate system. Nevertheless, I can get it to print the difference it finds, if that will appease you ;)

Reply:
I don't care much about the actual number, but the order of magnitude (if > 1m2). Otherwise, we can safely say: `# < 1 squaremeter` and everything can stop caring.

Reply:
So, looking at the case where it happens in Scotland (in my tests), there is *no* difference in area after buffering. In that instance, a 1m2 difference would be picked up at a ~1e-11 relative tolerance. I could set up the relative tolerance to be tighter, to vaguely represent this 1m2 limit, perhaps at ~1e-10? Setting an absolute difference would then need to handle coordinate reference systems, so I'd avoid that here if possible.

Reply:
You are of course right that area calculation is difficult with WGS84 (we should really move to EPSG3035).

My point is that this check is meaningless if one doesn't understand the tolerance. The only thing I know after this line is that some number A is smaller than another (small) number B. So A is small, but what means small?

I also don't have a quick fix here.

Reply:
well, no, after this line you know that two numbers are very similar, rather than one being any smaller than another. Indeed, the numbers could be identical, we just don't actually care for them to necessarily be identical. Rather, they should be 'close enough'. This comparison of 'close enough' is independent of the coordinate reference system and I think that a relative tolerance of 1e-10 would be more than sufficient to catch issues. If this means that it is >>1m2 across the whole continent, that is still negligible on the scale of the system, as < 1m2 is negligible on the scale of a Scottish island. 

Perhaps we revisit this with some concrete limit once we've moved to EPSG:3035 across the board?

Reply:
moving this discussion out of the PR and into #14

Reply:
Actually, this whole discussion is misguided by me not knowing that there is a relative tolerance in `np.isclose`. Therefore, I suggest to make this explicit, as in `np.isclose(a, b, rel_tol=0.001)`. In that case, we know the error is < 0.1% and irrelevant.

Reply:
Yup, it's known as `rtol` as I mentioned and I've set it to be much more strict: 1e-10!